plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -2905,12 +2912,41 @@ public boolean delegateMaterializedViewRefreshToConnector(ConnectorSession sessi
     }
 
     @Override
-    public ConnectorInsertTableHandle beginRefreshMaterializedView(ConnectorSession session, ConnectorTableHandle tableHandle, List<ConnectorTableHandle> sourceTableHandles, RetryMode retryMode)
+    public ConnectorInsertTableHandle beginRefreshMaterializedView(ConnectorSession session, ConnectorTableHandle tableHandle, List<ConnectorTableHandle> sourceTableHandles, RetryMode retryMode, RefreshType refreshType)
     {
+        checkState(fromSnapshotsForRefresh.isEmpty(), "Snapshot map must be empty at the start of MV refresh operation.");
         IcebergTableHandle table = (IcebergTableHandle) tableHandle;
         Table icebergTable = catalog.loadTable(session, table.getSchemaTableName());
         beginTransaction(icebergTable);
 
+        Optional<String> dependencies = Optional.ofNullable(icebergTable.currentSnapshot())
+                .map(Snapshot::summary)
+                .map(summary -> summary.get(DEPENDS_ON_TABLES));
+
+        boolean shouldUseIncremental = isIncrementalRefreshEnabled(session)
+                && refreshType == RefreshType.INCREMENTAL
+                // there is a single source table
+                && sourceTableHandles.size() == 1
+                // and source table is an Iceberg table
+                && getOnlyElement(sourceTableHandles) instanceof IcebergTableHandle
```

Comment:
Is it a problem to have the source table from a different catalog? if it's an iceberg table we would still accurately record its fromSnapshot in MV snapshot summary, no?

Reply:
I think I see your point around the namespacing of DB objects, since we only record the info as `schema.table=snapshotId`. However, as long as the MV doesn't have two source tables from two different catalogs with the same schema/table names, that wouldn't cause any problems. Also, right now we restrict the scope of incremental refresh for single-source table cases. 

To support colliding schema/table names too in the future, we would have to 1) change the snapshot summary logic to write `catalog.schema.table=snapshotId` and change the fromSnapshot map to incorporate the catalog too into its keys